Multimodality/sft extension

RaphaelKreft opened the pull request:
Implementation of a SFT Dataset for SFT Training in Megatron. Developed originally for Visual Instruction Tuning. 

-  As the gpt_dataset it uses an IndexedDataset as Low-Level Dataset.
-  From there it loads pre-tokenized sft data, then masks the user prompts (in our case including image tokens).
-  Per training sample: loads a single sample from indexed dataset, then pads to maximum sequence length

### Added
- `--sft` cli argument (when given, use the new SFTIndexedDataset)
- `SFTIndexedDataset` loading and preparing pre-tokenized sft data

### Questions / Todos

- [ ] Add dynamic loading of "begin of user prompt" and "end-of-turn" sequence (could be part of tokenizer, is currently hard-coded and thus only works for LLama3 Vision Model Chat templates)
- [ ] Implement option to mask loss on all special tokens (BOS, EOD, EOS, SFT-Related special tokens)??

RaphaelKreft committed:
add sft argument and preliminary implementation of SFTIndexedDataset

RaphaelKreft committed:
make SFTIndexedDataset align with GPTDataset.

RaphaelKreft committed:
Implement masking of user sequences

Commit:
Debug SFTIndexed Dataset

RaphaelKreft committed:
add masking for attention rows for padding tokens

Commit:
Remove special tokens from user start/end sequence

RaphaelKreft committed:
end truncated samples with eod token

RaphaelKreft committed:
Add option to mask special token(sequences) in sft dataset and improve efficiency of loss-mask creation.

Commit:
debugged sdt and tokenizer. added options to not mask image tokens

Alvorecer721 commented on megatron/core/datasets/sft_dataset.py:
```diff
@@ -419,3 +417,135 @@ def get_matching_mask(sequence, query: torch.Tensor, only_begin:bool=True):
     return matches
 
 
+def get_matching_mask_by_start_end(tokens, begin_seq: torch.Tensor, end_seq: torch.Tensor):
```
## CRUCIAL BUG: We're Searching in the Wrong Place

Let’s make this concrete with a simple example.

```
Sequence:  a b c d e

Tokens:    a b c d
Labels:    b c d e
```

Each `loss_mask[i]` controls whether we train on predicting `labels[i]` from `tokens[i]`, i.e.:

* `loss_mask[0]`: a → b
* `loss_mask[1]`: b → c
* `loss_mask[2]`: c → d
* `loss_mask[3]`: d → e

---

#### The Problem

Say we want to **mask the prediction of `'c'`** (don’t train the model to predict `'c'`).

**Current (incorrect) logic:**

* We search for `'c'` in **tokens** → find index 2
* Set `loss_mask[2] = 0`

Result:

```
a → b   (trained)
b → c   (trained)   ❌ we wanted to mask this one
c → d   (masked)
d → e   (trained)
```

This disables the *c→d* prediction instead of *b→c*.

---

#### Correct logic

We should search for `'c'` in **labels**, not tokens:

* `'c'` is at label index 1
* Set `loss_mask[1] = 0`

Result:

```
a → b   (trained)
b → c   (masked)    ✅ correct
c → d   (trained)
d → e   (trained)
```

RaphaelKreft replied:
I now renamed the arguments to the sequence matching methods (to "data" and "sequence")  and more importantly give them the labels to calculate the mask.

Alvorecer721 commented on megatron/core/datasets/sft_dataset.py:
```diff
@@ -419,3 +417,135 @@ def get_matching_mask(sequence, query: torch.Tensor, only_begin:bool=True):
+    end_len = len(end_seq)
+
+    if 0 < begin_len <= len(tokens):
+        matches_begin = get_matching_mask(tokens, begin_seq, only_begin=True)
```
Same applies here, get_matching_mask should work on the labels rather than tokens.

Imagine the sequence: `"<|start_header_id|>Assistant<|end_header_id|> Hi"`

We want to enable the loss from <|end_header_id|> → Hi, as the model should learn to predict “Hi” from <|end_header_id|>.

Alvorecer721 commented on megatron/core/datasets/sft_dataset.py:
```diff
@@ -0,0 +1,551 @@
+                "position_ids": position_ids,
+            }
+
+    def _get_ltor_masks_and_position_ids(self, tokens,
```
I would prefer using `data` as in the original `_get_ltor_masks_and_position_ids` against `tokens` here, in the original implementation tokens are passed here to prevent the generation after eod token in `tokens` during pretraining, for SFT maybe we should pass in `labels` instead.

Alvorecer721 commented on megatron/core/datasets/sft_dataset.py:
```diff
@@ -0,0 +1,551 @@
+        if not only_begin:
+            matches_float = matches.float().unsqueeze(0).unsqueeze(0)  # (1, 1, N)
+            kernel = torch.ones(1, 1, query_len, device=sequence.device)
+            expanded = F.conv1d(matches_float, kernel, padding=query_len - 1)
```
Do you think using `convid` here for padding is a bit overkill? @TJ-Solergibert

TJ-Solergibert replied:
Well, it can be, but as far as 1. Doesn't hurts performance and 2. Everyone is comfortable with it, it's fine. 

Keep in mind that this function is performed 1. In the CPU 2. While the GPU is processing the previous batch, so as far as you don't hit any CPU OOM error and you are not bottlenecked by the DataLoader you are good. To check for the later just compare the throughput when using mock data.

RaphaelKreft committed:
fix: use labels to calculate values for loss mask NOT tokens

RaphaelKreft committed:
implement right padding, add debug flag, remove goldfish loss, cleanup

RaphaelKreft committed:
remove caching as it potentially can cause issues. Add cmd arg for plw

Alvorecer721 committed:
Add tokenizer properties for pre-tokenized SFT sequences

Alvorecer721 committed:
Use pre-tokenized sequences from tokenizer config

- Load sequences from tokenizer properties instead of tokenizing at runtime
- Pre-compute token sequences as tensors in __init__
- Use .to() instead of torch.tensor() in hot path for efficiency
- Reduces overhead during training data loading

RaphaelKreft committed:
add support for plw in sft_dataset.py.

Commit:
fix: add sft_plw to gpt dataset config properly

RaphaelKreft committed:
add assistant loss logging for sft (untested)

RaphaelKreft committed:
improve assistant loss logging to work with plw=1 (untested)

Commit:
fix assistant mask not passed though callchain correctly

RaphaelKreft committed:
implement sample-packing for sft dataset (untested)

RaphaelKreft committed:
add and use python-version of build_packed_idx method in sft_dataset for now (untested)

Commit:
Fix minor issues to make packing training launch and compile

RaphaelKreft committed:
add sft init script that can init sft dataset once to print packed samples and then exit. Remove dummy packing arg and code from main code. (untested)

RaphaelKreft committed:
log packing statistics also if loaded from cache

Commit:
remove unecessary prints from init sft script. Sft dset prints sttistics itself.

RaphaelKreft committed:
Add --final-checkpoint arg, that enables storing a final checkpoint regardless of checkpoint interval

RaphaelKreft committed:
Add option to skip skip margin samples, to exactly reach target number of samples

RaphaelKreft committed:
Add option to skip skip margin samples, to exactly reach target number of samples

Commit:
Minor fixes to sft-dataset

RaphaelKreft committed:
re-enable forward pre-hook before final checkpoint

RaphaelKreft committed:
activate c++ index building helper, deprecate python one

RaphaelKreft committed:
remove not-mask img tokens, add sample averaging and multi-epoch support to sft-dataset.

Commit:
use correct shuffle in single doc idx building

RaphaelKreft committed:
extend key conf attributes of sft_dataset to account for packing

RaphaelKreft committed:
cleanup old python index building implementation. Limit sample index to samples needed.

RaphaelKreft committed:
store correct size of sample index and mitigate NaN when equalizing sequence loss.

RaphaelKreft committed:
move zero-out of loss-mask into respective method

RaphaelKreft committed:
fix attempt NaN loss

RaphaelKreft committed:
sft-dataset: obtain sample boundaries during low level sample loading not form eod indices

RaphaelKreft committed:
add load loss mask from disk option - cleanup sft dataset

Commit:
minor fixes to sft_dataset.py after refactor

RaphaelKreft committed:
Always compute assistant loss mask for sft. Add support for separate assistant/user end tokens. Sft masking optimization: only unmask assistant parts.

RaphaelKreft committed:
Fix: positional-ids and attention mask wrong computed for SFT dataset (offset by 1)

RaphaelKreft committed:
add script to create weighted data configs to support multimodal fixed img ratio

Commit:
remove return of empty assitant mask from gpt_dataset.py to fix collate

Commit:
fix: extended model checkpoint stores correct vocabulary size
